segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/IdempotentStorageBase.java

```diff
@@ -0,0 +1,103 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.impl.filesystem;
+
+import com.google.common.base.Preconditions;
+import io.pravega.common.Exceptions;
+import io.pravega.segmentstore.storage.Storage;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Base class for all the storage implementations that support idempotent write at an offset.
+ */
+public abstract class IdempotentStorageBase implements Storage {
```

Comment:
1. There are also some boilerplate code which is common and is extracted here. Do you have a good recommendation for the name of the class?
2. Do you have a good recommendation for the location?